go/arrow/ipc/message.go

```diff
@@ -0,0 +1,197 @@
+	MetadataV2 = MetadataVersion(flatbuf.MetadataVersionV2) // version for Arrow-0.2.0
+	MetadataV3 = MetadataVersion(flatbuf.MetadataVersionV3) // version for Arrow-0.3.0 to 0.7.1
+	MetadataV4 = MetadataVersion(flatbuf.MetadataVersionV4) // version for >= Arrow-0.8.0
+)
+
+func (m MetadataVersion) String() string {
+	if v, ok := flatbuf.EnumNamesMetadataVersion[int16(m)]; ok {
+		return v
+	}
+	return fmt.Sprintf("MetadataVersion(%d)", int16(m))
+}
+
+// MessageType represents the type of Message in an Arrow format.
+type MessageType flatbuf.MessageHeader
+
+const (
+	MessageNone            = MessageType(flatbuf.MessageHeaderNONE)
+	MessageSchema          = MessageType(flatbuf.MessageHeaderSchema)
+	MessageDictionaryBatch = MessageType(flatbuf.MessageHeaderDictionaryBatch)
+	MessageRecordBatch     = MessageType(flatbuf.MessageHeaderRecordBatch)
+	MessageTensor          = MessageType(flatbuf.MessageHeaderTensor)
+	MessageSparseTensor    = MessageType(flatbuf.MessageHeaderSparseTensor)
+)
+
+func (m MessageType) String() string {
+	if v, ok := flatbuf.EnumNamesMessageHeader[byte(m)]; ok {
+		return v
+	}
+	return fmt.Sprintf("MessageType(%d)", int(m))
+}
+
+const (
+	// maxNestingDepth is an arbitrary value to catch user mistakes.
+	// For deeply nested schemas, it is expected the user will indicate
+	// explicitly the maximum allowed recursion depth.
+	maxNestingDepth = 64
+)
+
+// Message is an IPC message, including metadata and body.
+type Message struct {
+	refCount int64
+	msg      *flatbuf.Message
+	meta     *memory.Buffer
+	body     *memory.Buffer
+}
+
+func NewMessage(meta, body *memory.Buffer) *Message {
+	meta.Retain()
+	body.Retain()
+	return &Message{
+		refCount: 1,
+		msg:      flatbuf.GetRootAsMessage(meta.Bytes(), 0),
+		meta:     meta,
+		body:     body,
+	}
+}
+
+// Retain increases the reference count by 1.
+// Retain may be called simultaneously from multiple goroutines.
+func (msg *Message) Retain() {
+	atomic.AddInt64(&msg.refCount, 1)
+}
+
+// Release decreases the reference count by 1.
+// Release may be called simultaneously from multiple goroutines.
+// When the reference count goes to zero, the memory is freed.
+func (msg *Message) Release() {
+	debug.Assert(atomic.LoadInt64(&msg.refCount) > 0, "too many releases")
+
+	if atomic.AddInt64(&msg.refCount, -1) == 0 {
+		msg.meta.Release()
+		msg.body.Release()
+		msg.msg = nil
+		msg.meta = nil
+		msg.body = nil
+	}
+}
+
+func (msg Message) Version() MetadataVersion {
+	return MetadataVersion(msg.msg.Version())
+}
+
+func (msg Message) Type() MessageType {
+	return MessageType(msg.msg.HeaderType())
+}
+
+func (msg Message) BodyLen() int64 {
+	return msg.msg.BodyLength()
+}
+
+// MessageReader reads messages from an io.Reader.
+type MessageReader struct {
+	r io.Reader
+
+	refCount int64
+	msg      *Message
+}
+
+// NewMessageReader returns a reader that reads messages from an input stream.
+func NewMessageReader(r io.Reader) (*MessageReader, error) {
```

Comment:
Why do we return an error here?

Reply:
a leftover from my initial pie-in-the-sky "design" (where I reasonned that as it was taking an io.Reader, it would read stuff out of it and then perhaps fail.)

but I think you're right.
I'll fix that. thanks!